File: cocos/cocos2d-x/cocos/platform/android/java/src/org/cocos2dx/custom/DownloadTask.java
package eggy.cocos2dx.custom;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Environment;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class DownloadTask {
    private static DownloadTask downloadUtil;
    private final OkHttpClient okHttpClient;

    public static DownloadTask get() {
        if (downloadUtil == null) {
            downloadUtil = new DownloadTask();
        }
        return downloadUtil;
    }

    private DownloadTask() {
        okHttpClient = new OkHttpClient();
    }

    /**
     * @param url      下载连接
     * @param saveDir  储存下载文件的SDCard目录
     * @param listener 下载监听
     */
    public void download(final String url, final String saveDir, final OnDownloadListener listener) {
        Request request = new Request.Builder().url(url).build();
        okHttpClient.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                listener.onDownloadFailed();
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                InputStream is = null;
                byte[] buf = new byte[2048];
                int len = 0;
                FileOutputStream fos = null;
                String savePath = isExistDir(saveDir);
                try {
                    is = response.body().byteStream();
                    long total = response.body().contentLength();
                    File saveFile = new File(savePath, getNameFromUrl(url));
                    fos = new FileOutputStream(saveFile);
                    long sum = 0;
                    while ((len = is.read(buf)) != -1) {
                        fos.write(buf, 0, len);
                        sum += len;
                        int progress = (int) (sum * 1.0f / total * 100);
                        listener.onDownloading(progress);
                    }
                    fos.flush();
                    listener.onDownloadSuccess(saveFile);
                } catch (Exception e) {
                    listener.onDownloadFailed();
                } finally {
                    try {
                        if (is != null)
                            is.close();
                    } catch (IOException e) {
                    }
                    try {
                        if (fos != null)
                            fos.close();
                    } catch (IOException e) {
                    }
                }
            }
        });
    }

    public void downloadJpeg(Context context, String url, String filename, OnDownloadListener listener) {
        Request request = new Request.Builder().url(url).build();
        okHttpClient.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                listener.onDownloadFailed();
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                InputStream is = null;
                FileOutputStream fos = null;
                try {
                    is = response.body().byteStream();
                    BitmapFactory.Options ops = new BitmapFactory.Options();
                    ops.inJustDecodeBounds = false;
                    Bitmap bmp = BitmapFactory.decodeStream(is, null, ops);
                    File file = new File(context.getFilesDir(), filename);
                    fos = new FileOutputStream(file);
                    bmp.compress(Bitmap.CompressFormat.JPEG, 80, fos);
                    fos.flush();
                    listener.onDownloadSuccess(file);
                } catch (Exception e) {
                    e.printStackTrace();
                    listener.onDownloadFailed();
                } finally {
                    if (is != null)
                        is.close();
                    if (fos != null)
                        fos.close();
                }
            }
        });
    }

    /**
     * @param saveDir
     * @return
     * @throws IOException 判断下载目录是否存在
     */
    private String isExistDir(String saveDir) throws IOException {
        File downloadFile = new File(Environment.getExternalStorageDirectory(), saveDir);
        if (!downloadFile.mkdirs()) {
            downloadFile.createNewFile();
        }
        String savePath = downloadFile.getAbsolutePath();
        return savePath;
    }

    public File getAppDir(String saveDir, String url) {
        return new File(Environment.getExternalStorageDirectory(), saveDir + "/" + getNameFromUrl(url));
    }

    /**
     * @param url
     * @return 从下载连接中解析出文件名
     */
    public String getNameFromUrl(String url) {
        return url.substring(url.lastIndexOf("/") + 1);
    }

    public interface OnDownloadListener {
        /**
         * 下载成功
         */
        void onDownloadSuccess(File saveFile);

        /**
         * @param progress 下载进度
         */
        void onDownloading(int progress);

        /**
         * 下载失败
         */
        void onDownloadFailed();
    }
}

File: lib-engine/src/main/kotlin/eggy/cocos2dx/custom/DownloadTask.kt
package eggy.cocos2dx.custom

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Environment
import eggy.util.AppGlobal
import eggy.util.IOUtil
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Request.Builder
import okhttp3.Response
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream

class DownloadTask private constructor() {
    private val okHttpClient: OkHttpClient = OkHttpClient()

    /**
     * @param url      下载连接
     * @param saveDir  储存下载文件的SDCard目录
     * @param listener 下载监听
     */
    fun download(url: String, saveDir: String, listener: OnDownloadListener) {
        val request: Request = Builder().url(url).build()
        okHttpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                listener.onDownloadFailed()
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                var `is`: InputStream? = null
                val buf = ByteArray(2048)
                var len: Int
                var fos: FileOutputStream? = null
                val savePath = getDownloadDir(saveDir)
                try {
                    `is` = response.body!!.byteStream()
                    val total = response.body!!.contentLength()
                    val saveFile = File(savePath, getNameFromUrl(url))
                    fos = FileOutputStream(saveFile)
                    var sum: Long = 0
                    while (`is`.read(buf).also { len = it } != -1) {
                        fos.write(buf, 0, len)
                        sum += len.toLong()
                        val progress = (sum * 1.0f / total * 100).toInt()
                        listener.onDownloading(progress)
                    }
                    fos.flush()
                    listener.onDownloadSuccess(saveFile)
                } catch (e: Exception) {
                    listener.onDownloadFailed()
                } finally {
                    IOUtil.close(`is`)
                    IOUtil.close(fos)
                }
            }
        })
    }

    fun downloadJpeg(
        context: Context,
        url: String?,
        filename: String?,
        listener: OnDownloadListener
    ) {
        val request: Request = Builder().url(url!!).build()
        okHttpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                listener.onDownloadFailed()
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                var `is`: InputStream? = null
                var fos: FileOutputStream? = null
                try {
                    `is` = response.body!!.byteStream()
                    val ops = BitmapFactory.Options()
                    ops.inJustDecodeBounds = false
                    val bmp = BitmapFactory.decodeStream(`is`, null, ops)
                    val file = File(context.filesDir, filename)
                    fos = FileOutputStream(file)
                    bmp!!.compress(Bitmap.CompressFormat.JPEG, 80, fos)
                    fos.flush()
                    listener.onDownloadSuccess(file)
                } catch (e: Exception) {
                    e.printStackTrace()
                    listener.onDownloadFailed()
                } finally {
                    IOUtil.close(`is`)
                    IOUtil.close(fos)
                }
            }
        })
    }

    private fun getDownloadDir(saveDir: String): File {
        val downloadFile = File(
            AppGlobal.application.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS),
            saveDir
        )
        if (!downloadFile.mkdirs()) {
            downloadFile.createNewFile()
        }
        return downloadFile
    }



    /**
     * @param url
     * @return 从下载连接中解析出文件名
     */
    fun getNameFromUrl(url: String): String {
        return url.substring(url.lastIndexOf("/") + 1)
    }

    interface OnDownloadListener {
        /**
         * 下载成功
         */
        fun onDownloadSuccess(saveFile: File?)

        /**
         * @param progress 下载进度
         */
        fun onDownloading(progress: Int)

        /**
         * 下载失败
         */
        fun onDownloadFailed()
    }

    companion object {
        private var downloadUtil: eggy.cocos2dx.custom.DownloadTask? = null
        fun get(): eggy.cocos2dx.custom.DownloadTask? {
            if (eggy.cocos2dx.custom.DownloadTask.Companion.downloadUtil == null) {
                eggy.cocos2dx.custom.DownloadTask.Companion.downloadUtil =
                    eggy.cocos2dx.custom.DownloadTask()
            }
            return eggy.cocos2dx.custom.DownloadTask.Companion.downloadUtil
        }
    }
}